src/db/file.rs

```diff
@@ -138,102 +138,96 @@ pub fn add_path_into_database<P: AsRef<Path>>(conn: &Connection,
-            };
+            } else {
+                mime.to_owned()
+            }
+        };
 
-            let content: Option<Vec<u8>> = if let Some(client) = &mut client {
-                let mut attempts = 0;
-                loop {
-                    let s3_res = client.put_object(PutObjectRequest {
+        if let Some(client) = s3_client() {
+            let bucket_path = bucket_path.clone();
+            let content = content.clone();
+            let mime = mime.clone();
+
+            futures.push(client
+                .put_object(PutObjectRequest {
+                    bucket: "rust-docs-rs".into(),
+                    key: bucket_path.clone(),
+                    body: Some(content.clone().into()),
+                    content_type: Some(mime.clone()),
+                    ..Default::default()
+                })
+                .map_err(move |e| {
+                    log::error!("failed to upload to {}: {:?}", bucket_path, e);
+                    // try two additional times (for a total of 3)
+                    client.put_object(PutObjectRequest {
                         bucket: "rust-docs-rs".into(),
                         key: bucket_path.clone(),
                         body: Some(content.clone().into()),
                         content_type: Some(mime.clone()),
                         ..Default::default()
-                    }).sync();
-                    attempts += 1;
-                    match s3_res {
-                        // we've successfully uploaded the content, so steal it;
-                        // we don't want to put it in the DB
-                        Ok(_) => break None,
-                        // Since s3 was configured, we want to panic on failure to upload.
-                        Err(e) => {
-                            log::error!("failed to upload to {}: {:?}", bucket_path, e);
-                            // Get a new client, in case the old one's connection is stale.
-                            // AWS will kill our connection if it's alive for too long; this avoids
-                            // that preventing us from building the crate entirely.
-                            *client = s3_client().unwrap();
-                            if attempts > 3 {
-                                panic!("failed to upload 3 times, exiting");
-                            } else {
-                                continue;
-                            }
-                        },
-                    }
-                }
-            } else {
-                Some(content.clone().into())
-            };
-
-            file_list_with_mimes.push((mime.clone(), file_path.clone()));
-
-            (
-                bucket_path,
-                content,
-                mime,
-            )
-        };
-
-        // If AWS credentials are configured, don't insert/update the database
-        if client.is_none() {
+                    })
+                    .into_stream()
```

Comment:
This won't do what we want unfortunately. I think it should be fine to regress to just retrying once though so we can just skip the streaming here I think and just retry the PUT.

Reply:
Dang I was trying to get fancy and bundle all that behavior into a Future :)